### PR TITLE
[ML] Allow on-the-fly adjustment of the number of threads in use

### PR DESCRIPTION
**Pull request by tveasey**
For PyTorch model inference we want to be able to control the number of threads we assign to parallel calls to forward on-the-fly to adjust the number of cores each model consumes on a node as cluster wide tasks are added or removed. This lays the groundwork by adding a dynamic setting for the number of threads the pool will use. The idea is we always start the process thread pool with the hardware concurrency (divided by the number of threads we give libtorch), but then adjust the threads actually used by control message. Unused threads will simply be idle waiting to pop an empty queue so are effectively free.

cc @dimitris-athanasiou.

### CODE REVIEW

File: lib/core/CStaticThreadPool.cc [hunk 1]
```diff
@@ -40,9 +41,14 @@ CStaticThreadPool::~CStaticThreadPool() {
     this->shutdown();
 }
 
+void CStaticThreadPool::numberThreadsInUse(std::size_t threads) {
+    threads = std::max(std::min(threads, m_Pool.size()), 1UL);
```
Comment (droberts195) [inline lib/core/CStaticThreadPool.cc:45]:
```suggestion
    threads = std::max(std::min(threads, m_Pool.size()), std::size_t{1});
```

I think `size_t` is `unsigned long long` on some platforms, so the `UL` suffix might not be portable.

File: lib/core/unittest/CStaticThreadPoolTest.cc [hunk 2]
```diff
@@ -18,7 +18,9 @@
 
 #include <atomic>
 #include <chrono>
+#include <mutex>
 #include <thread>
+#include <unordered_set>
```
Comment (droberts195) [inline lib/core/unittest/CStaticThreadPoolTest.cc:23]: We're still using `boost::unordered_set` everywhere else.  I think it's best we stick with one across the codebase.  Using `std` here will just encourage mixing the two.

Comment (tveasey) [reply lib/core/unittest/CStaticThreadPoolTest.cc]: See [f047acd](https://github.com/elastic/ml-cpp/pull/2232/commits/f047acdece21675beec11357850a4b73d3defdd6).

File: lib/core/CStaticThreadPool.cc (hunk 1 above)
Comment (tveasey) [reply lib/core/CStaticThreadPool.cc]: See [f047acd](https://github.com/elastic/ml-cpp/pull/2232/commits/f047acdece21675beec11357850a4b73d3defdd6).

**Review by droberts195 — APPROVED** [review]
LGTM

File: lib/core/unittest/CStaticThreadPoolTest.cc [hunk 3]
```diff
@@ -202,7 +204,46 @@ BOOST_AUTO_TEST_CASE(testWithExceptions) {
...
+                executionThreads.insert(std::this_thread::get_id());
+            });
+        }
+
+        for (;;) {
+            std::scoped_lock<std::mutex> lock{mutex};
+            if (numberProcessedTasks == 200) {
+                LOG_DEBUG(<< "# threads used = " << executionThreads.size());
+                // A subset of threads can steal all the work.
+                BOOST_REQUIRE(numberThreadsInUse <= executionThreads.size());
```
Comment (dimitris-athanasiou) [inline lib/core/unittest/CStaticThreadPoolTest.cc:240]: Shouldn't we be asserting that `executionThreads.size() <= numberThreadsInUse` instead?

Comment (tveasey) [reply lib/core/unittest/CStaticThreadPoolTest.cc:240]: We should indeed!

**Review by dimitris-athanasiou — COMMENTED** [review]
LGTM

**Pull request merged**